// @framer/colin.loupe-checkbox-afternoon/code/Checkbox.tsx
import * as React from "react"
import {
    Frame,
    useCycle,
    addPropertyControls,
    PropertyControls,
    ControlType,
} from "framer"
//@ts-ignore
import { colors } from "@framer/aroagb.loupe-colors/code/canvas"
import { Checked } from "./canvas"

// import { Checkbox } from "./canvas"

export function Checkbox(props) {
    const [active, inactive] = useCycle({ opacity: 0 }, { opacity: 1 })

    return (
        <>
            <Frame
                style={{
                    backgroundColor: "rgba(209, 209, 209, 0.16)",
                    border: "2px solid #d1d1d1",
                    borderRadius: "4px",
                    overflow: "visible",
                    width: 24,
                    height: 24,
                }}
                whileHover={{
                    backgroundColor: "rgba(1, 85, 254, 0.08)",
                    border: "2px solid #0155FE",
                    borderRadius: "4px",
                    boxShadow: "0px 2px 6px 0px rgba(1, 85, 254, 0.16)",
                    overflow: "visible",
                }}
                onClick={{
                    background: "#0055ff",
                    border: "2px solid #0155FE",
                    borderRadius: "4px",
                    overflow: "visible",
                }}
                // animate={active}
                // onTap={inactive}
            ></Frame>
            <Checked />
        </>
    )
}

// Change the width and height here to fit your component
Checkbox.defaultProps = {
    width: 24,
    height: 24,
}
